### PR TITLE
implement tslibs.arraylike

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -37,6 +37,9 @@ from cython cimport Py_ssize_t
 import pytz
 UTC = pytz.utc
 
+from tslibs.arraylike import (  # noqa:F841
+    format_array_from_datetime, array_to_datetime, array_with_unit_to_datetime,
```
Comment (jreback) [inline pandas/_libs/tslib.pyx]: isn’t the point to not import back here, rather import directly to where it’s needed

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslib.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: Sure.  For the first pass I like to keep the diff small, but now I'll update.

Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx]: Actually on second thought there are a _lot_ of usages of these I'd rather not update.  I think the thing to do is leave this where it is for now and put it in the `tslibs` namespace a few PRs from now when we finish transitioning `tslib`

File: pandas/_libs/tslibs/arraylike.pyx [hunk 2]
```diff
@@ -0,0 +1,717 @@
+# -*- coding: utf-8 -*-
+
+cimport numpy as cnp
+from numpy cimport int64_t, ndarray, float64_t
+import numpy as np
```
Comment (jreback) [inline pandas/_libs/tslibs/arraylike.pyx]: so some of this should be put in conversion. or is there a reason you are not doing that? i would like to shy away from creating a new module unless it has a clear purpose. can you move what is logical first.

File: pandas/_libs/tslibs/arraylike.pyx [hunk 3]
```diff
@@ -0,0 +1,717 @@
+# -*- coding: utf-8 -*-
+
```
Comment (jorisvandenbossche) [inline pandas/_libs/tslibs/arraylike.pyx]: if you end up creating this file, can you take the opportunity of the refactor to add a module docstring here to better document what / what type of functionality is included in this file? 

But I agree with Jeff that from a quick view some of those functions seems to be about parsing / conversion?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

**Comment by jbrockmendel** [comment]
Woops, closed the wrong one.

**Comment by jbrockmendel** [comment]
Closing as non-actionable for now

**Pull request closed without merging**